plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,498 @@
+		b := r.Replace(err.Error())
+		s.Log.Errorf("An error occured %v", b)
+		return false
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		s.Log.Errorf("Error while reading the response bytes: %v", err)
+		return false
+	}
+
+	err = json.Unmarshal(body, &s.token)
+	if err != nil {
+		s.Log.Errorf("An error occured %v", err)
+		return false
+	}
+
+	return true
+}
+
+// createSubscription uses the official 'ctrlX Data Layer API' to create the sse subscription
+// Requests are sent in two seconds interval to reduce load on data layer.
+func (s *ctrlXDataLayerSSE) createSubscription(sub *subscription) string {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	time.Sleep(time.Duration(defaultCreateSubscriptionInterval))
+
+	sseURL := s.Hostname + "/automation/api/v2/events"
+
+	pl := sub.payload()
+	p, _ := json.Marshal(pl)
+
+	requestBody := bytes.NewBuffer(p)
+	req, err := http.NewRequest("POST", sseURL, requestBody)
+	if err != nil {
+		s.Log.Errorf("Failed to create post request to create sse subscription %v, error: %v", sub.index, err)
+		return ""
+	}
+
+	req.Header.Add("Authorization", s.token.String())
+
+	resp, err := s.connection.Do(req)
+	if err != nil {
+		s.Log.Errorf("Failed to do post request to create sse subscription %v, error: %v", sub.index, err)
+		return ""
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 && resp.StatusCode != 201 {
+		s.Log.Errorf("Failed to create sse subscription %v, status: %v", sub.index, resp.Status)
+		return ""
+	}
+
+	return sseURL + "/" + pl.Properties.ID
+}
+
+// reconnect is called when the connection is broken
+// reconnect starts restablishing the connection
+func (s *ctrlXDataLayerSSE) reconnect(sub subscription) {
+	go func() {
+		s.Log.Infof("Subscription %v is down, reconnect is starting", sub.index)
+		for {
+			time.Sleep(time.Duration(defaultReconnectInterval))
+
+			if !s.setAuthToken() {
+				continue
+			}
+
+			subURL := s.createSubscription(&sub)
+			if subURL == "" {
+				continue
+			}
+
+			client := newClient(subURL, sub, s.Log)
+			if s.autoReconnect {
+				client.onDisconnect(s.reconnect)
+			}
+			client.Headers["Authorization"] = s.token.String()
+			client.Connection.Transport = httpTransport(s.InsecureSkipVerify)
+			err := client.requestData(s, sub.Measurement)
+			if err != nil {
+				continue
+			}
+
+			s.Log.Infof("Subscription %v is reconnected", sub.index)
+			return
+		}
+	}()
+}
+
+// addMetric writes sse metric into accumulator
+func (s *ctrlXDataLayerSSE) addMetric(se *sSEEvent, measurement string) {
+	m := metric{}
+	switch se.Event {
+	case "update":
+		// Received an updated value, that we translate into a metric
+		_ = json.Unmarshal([]byte(se.Data), &m)
+		t := convertTimestamp2UnixTime(m.Timestamp)
+		if s.acc != nil {
+			s.wg.Add(1)
+			s.acc.AddFields(measurement, map[string]interface{}{"value": m.Value}, map[string]string{"node": m.Node, "type": m.Type, "host": s.Hostname}, t)
+			s.wg.Done()
```

Comment:
What are you trying to achieve with the wait-group here?

Reply:
That's necessary to synchronize reconnect or cancel of subscriptions when the service is stopped.